features/environment.py

```diff
@@ -704,6 +669,57 @@ def create_instance_with_uat_installed(
     return inst
 
 
+def _get_user_data_for_instance(
```

Comment:
an out-of-scope comment: I think using user-data for these various setups is actually complicating our code. I think it may be more straightforward to start the lxd container, and then run a python function that runs commands and writes files for the given set up.

This would also allow for an easier way to do tests that require setup _before_ upgrading ua to the version under test. (we normally just do these as one-off sru scripts, but ideally they'd be run as integration tests to avoid regressions)

I may explore this at some point

Reply:
That's a great Idea - @blackboxsw is there a strong reason we set the PPAs through cloud-init instead of configuring it after instance creation, right before installing u-a-t?

Reply:
@renanrodrigo sorry to have missed this question. No strong reason, cloud-init happens to order apt configuration as being right right before any ubuntu_advantage #cloud-config directives, So we knew apt setup would be completed by cloud-init prior to reacting to any image-creation or test-specific ubuntu_advantage #cloud-config directives.  I don't see a problem with teasing out that aspect of user-data setup as long as you know the PPA is already setup in image creation before installing any test-related packages.